<a href="https://colab.research.google.com/github/sagar2582/AirBnB-Hotel-Booking-Analysis/blob/main/World_Bank_Binary_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import matplotlib.pyplot as plt 
import numpy as np 
import os # accessing directory structure
import pandas as pd 

In [ ]:
# Connecting to Drive to import the dataset

from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [ ]:
df = pd.read_csv('/content/drive/My Drive/Taiyo/world_bank_dataset.csv')

In [ ]:
# Creating a copy of our dataset for futher use

df_copy = df.copy()

In [ ]:
# First look
df.head()

id                    regionname  \
0  P151169  Middle East and North Africa   
1  P154943         East Asia and Pacific   
2  P161382         East Asia and Pacific   
3  P163922         East Asia and Pacific   
4  P169279         East Asia and Pacific   

                                         countryname prodline  \
0              Kingdom of Morocco;Kingdom of Morocco       PE   
1                  Kingdom of Tonga;Kingdom of Tonga       PE   
2  Republic of the Marshall Islands;Republic of t...       PE   
3  Federated States of Micronesia;Federated State...       PE   
4                                        Samoa;Samoa       PE   

                   lendinginstr lendinginstrtype envassesmentcategorycode  \
0  Investment Project Financing               IN                        B   
1  Investment Project Financing               IN                        C   
2  Investment Project Financing               IN                        B   
3  Investment Project Financing               IN                        B   
4  Investment Project Financing               IN                      NaN   

  supplementprojectflg productlinetype projectstatusdisplay  ... mjtheme3name  \
0                    N               L               Active  ...          NaN   
1                    N               L               Active  ...          NaN   
2                    N               L               Active  ...          NaN   
3                    N               L               Active  ...          NaN   
4                    Y               L               Active  ...          NaN   

  mjtheme4name mjtheme5name location GeoLocID GeoLocName Latitude Longitude  \
0          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
1          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
2          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
3          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
4          NaN          NaN      NaN      NaN        NaN      NaN       NaN   

  Country Unnamed: 56  
0     NaN         NaN  
1     NaN         NaN  
2     NaN         NaN  
3     NaN         NaN  
4     NaN         NaN  

[5 rows x 57 columns]

In [ ]:
df.tail()

id                   regionname  \
18836  P006578  Latin America and Caribbean   
18837  P037451      Europe and Central Asia   
18838  P037362      Europe and Central Asia   
18839  P037452      Europe and Central Asia   
18840  P037383      Europe and Central Asia   

                                             countryname prodline  \
18836                Republic of Chile;Republic of Chile       PE   
18837  Grand Duchy of Luxembourg;Grand Duchy of Luxem...       PE   
18838              Kingdom of Denmark;Kingdom of Denmark       PE   
18839  Kingdom of the Netherlands;Kingdom of the Neth...       PE   
18840                    French Republic;French Republic       PE   

                     lendinginstr lendinginstrtype envassesmentcategorycode  \
18836    Specific Investment Loan               IN                      NaN   
18837  Structural Adjustment Loan               AD                      NaN   
18838  Structural Adjustment Loan               AD                      NaN   
18839  Structural Adjustment Loan               AD                      NaN   
18840  Structural Adjustment Loan               AD                      NaN   

      supplementprojectflg productlinetype projectstatusdisplay  ...  \
18836                    N               L               Closed  ...   
18837                    N               L               Closed  ...   
18838                    N               L               Closed  ...   
18839                    N               L               Closed  ...   
18840                    N               L               Closed  ...   

      mjtheme3name mjtheme4name mjtheme5name location GeoLocID GeoLocName  \
18836          NaN          NaN          NaN      NaN      NaN        NaN   
18837          NaN          NaN          NaN      NaN      NaN        NaN   
18838          NaN          NaN          NaN      NaN      NaN        NaN   
18839          NaN          NaN          NaN      NaN      NaN        NaN   
18840          NaN          NaN          NaN      NaN      NaN        NaN   

      Latitude Longitude Country Unnamed: 56  
18836      NaN       NaN     NaN         NaN  
18837      NaN       NaN     NaN         NaN  
18838      NaN       NaN     NaN         NaN  
18839      NaN       NaN     NaN         NaN  
18840      NaN       NaN     NaN         NaN  

[5 rows x 57 columns]

In [ ]:
df.shape

(18841, 57)

# Dataset Information
The shape of our dataset is (18841, 57)

id --> unique id-number for every project i.e., 18841 unique id's

regionname --> points towards the particular region of the world where the project belongs from

countryname --> name of the country

prodline --> product line - PE or RE (No idea what's this. I let the algo work on it)

lendinginstr --> different investment purposes

lendinginstrtype --> type of investment

envassesmentcategorycode --> investment category code

supplementprojectflg --> (No idea what's this) binary categoriacal column

productlinetype --> Single variable column. No use

projectstatusdisplay --> current live status

**status** --> overall status (Target Variable)

and more .....

## I see a lot of columns with a lot of NULL values, so let's start with cleaning our data and making it ready for visualization

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18841 entries, 0 to 18840
Data columns (total 57 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        18841 non-null  object 
 1   regionname                18841 non-null  object 
 2   countryname               18841 non-null  object 
 3   prodline                  18841 non-null  object 
 4   lendinginstr              18580 non-null  object 
 5   lendinginstrtype          18580 non-null  object 
 6   envassesmentcategorycode  12731 non-null  object 
 7   supplementprojectflg      18792 non-null  object 
 8   productlinetype           18841 non-null  object 
 9   projectstatusdisplay      18839 non-null  object 
 10  status                    18839 non-null  object 
 11  project_name              18841 non-null  object 
 12  boardapprovaldate         17172 non-null  object 
 13  board_approval_month      17172 non-null  object 
 14  closin

In [ ]:
df.describe(include='all')

id regionname                          countryname prodline  \
count     18841      18841                                18841    18841   
unique    18841          7                                  206       10   
top     P151169     Africa  Republic of India;Republic of India       PE   
freq          1       5900                                  806    15119   
mean        NaN        NaN                                  NaN      NaN   
std         NaN        NaN                                  NaN      NaN   
min         NaN        NaN                                  NaN      NaN   
25%         NaN        NaN                                  NaN      NaN   
50%         NaN        NaN                                  NaN      NaN   
75%         NaN        NaN                                  NaN      NaN   
max         NaN        NaN                                  NaN      NaN   

                    lendinginstr lendinginstrtype envassesmentcategorycode  \
count                      18580            18580                    12731   
unique                        18                4                        6   
top     Specific Investment Loan               IN                        B   
freq                        7702            16399                     5738   
mean                         NaN              NaN                      NaN   
std                          NaN              NaN                      NaN   
min                          NaN              NaN                      NaN   
25%                          NaN              NaN                      NaN   
50%                          NaN              NaN                      NaN   
75%                          NaN              NaN                      NaN   
max                          NaN              NaN                      NaN   

       supplementprojectflg productlinetype projectstatusdisplay  ...  \
count                 18792           18841                18839  ...   
unique                    2               1                    4  ...   
top                       N               L               Closed  ...   
freq                  16831           18841                14625  ...   
mean                    NaN             NaN                  NaN  ...   
std                     NaN             NaN                  NaN  ...   
min                     NaN             NaN                  NaN  ...   
25%                     NaN             NaN                  NaN  ...   
50%                     NaN             NaN                  NaN  ...   
75%                     NaN             NaN                  NaN  ...   
max                     NaN             NaN                  NaN  ...   

       mjtheme3name mjtheme4name mjtheme5name  \
count           0.0          0.0          0.0   
unique          NaN          NaN          NaN   
top             NaN          NaN          NaN   
freq            NaN          NaN          NaN   
mean            NaN          NaN          NaN   
std             NaN          NaN          NaN   
min             NaN          NaN          NaN   
25%             NaN          NaN          NaN   
50%             NaN          NaN          NaN   
75%             NaN          NaN          NaN   
max             NaN          NaN          NaN   

                                                 location  \
count                                                4569   
unique                                               4085   
top     0000281132!$!Gaza Strip!$!31.41667!$!34.333328...   
freq                                                   13   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                       

### Handling missing Data

In [ ]:
# Missing Value Count Function

def show_missing():
    missing = df.columns[df.isnull().any()].tolist()
    return missing

# Missing data counts and percentage
print('Missing Data Count')
print(df[show_missing()].isnull().sum().sort_values(ascending = False))
print('--'*50)
print('Missing Data Percentage')
print(round(df[show_missing()].isnull().sum().sort_values(ascending = False)/len(df)*100,2))

Missing Data Count
mjsector4                   18841
majorsector_percent         18841
mjtheme5name                18841
mjtheme4name                18841
mjtheme3name                18841
mjtheme2name                18841
mjtheme1name                18841
theme                       18841
mjsector5                   18841
mjsector3                   18841
mjsector2                   18841
mjsector1                   18841
Unnamed: 56                 18841
projectdoc                  18841
sector5                     16071
theme5                      15906
Country                     14390
sector4                     14361
GeoLocName                  14273
Longitude                   14272
Latitude                    14272
GeoLocID                    14272
location                    14272
theme4                      13594
sector3                     12169
financier                   11642
theme3                      11333
sector2                      9031
theme2                       

## Removing Features with over 40% missing values

In [ ]:
# Now working on df_copy

df_copy = df_copy.drop(['mjsector4','mjtheme5name','mjtheme4name','mjtheme3name','mjtheme2name','mjtheme1name','mjsector5','mjsector3',
                        'mjsector2','mjsector1','Unnamed: 56','sector5','theme5','Country','sector4','GeoLocName','Longitude','Latitude','GeoLocID',
                        'location','theme4','sector3','financier','theme3','sector2','theme2','projectdoc ','theme ','majorsector_percent '], axis = 1)

In [ ]:
# Missing Value Count Function

def show_missing():
    missing = df_copy.columns[df_copy.isnull().any()].tolist()
    return missing

# Missing data counts and percentage
print('Missing Data Count')
print(df_copy[show_missing()].isnull().sum().sort_values(ascending = False))
print('--'*50)
print('Missing Data Percentage')
print(round(df_copy[show_missing()].isnull().sum().sort_values(ascending = False)/len(df_copy)*100,2))

Missing Data Count
goal                        6653
impagency                   6553
borrower                    6369
envassesmentcategorycode    6110
closingdate                 3545
boardapprovaldate           1669
board_approval_month        1669
sector                       776
mjsector                     776
lendinginstr                 261
lendinginstrtype             261
lendprojectcost              132
supplementprojectflg          49
projectstatusdisplay           2
status                         2
dtype: int64
----------------------------------------------------------------------------------------------------
Missing Data Percentage
goal                        35.31
impagency                   34.78
borrower                    33.80
envassesmentcategorycode    32.43
closingdate                 18.82
boardapprovaldate            8.86
board_approval_month         8.86
sector                       4.12
mjsector                     4.12
lendinginstr                 1.39
lendingi

In [ ]:
df_copy.columns

Index(['id', 'regionname', 'countryname', 'prodline', 'lendinginstr',
       'lendinginstrtype', 'envassesmentcategorycode', 'supplementprojectflg',
       'productlinetype', 'projectstatusdisplay', 'status', 'project_name',
       'boardapprovaldate', 'board_approval_month', 'closingdate',
       'lendprojectcost', 'ibrdcommamt', 'idacommamt', 'totalamt', 'grantamt',
       'borrower', 'impagency', 'url', 'projectdoc ', 'majorsector_percent ',
       'sector1', 'sector', 'mjsector', 'theme1', 'theme ', 'goal'],
      dtype='object')

Further Inspecting each column and checking the relevance for our result

## **goal**

In [ ]:
df_copy['goal'].unique()

array(['Corporate Advocacy Priorities;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities|Millennium Development Goals;CAP|MDG;Corporate Advocacy Priorities;CAP',
       nan,
       'Global Public Goods Priorities|Millennium Development Goals;Global Public Goods Priorities|Millennium Development Goals;GPG|MDG;Corporate Advocacy Priorities;CAP;Millennium Development Goals|Global Public Goods Priorities;MDG|GPG',
       ...,
       'Corporate Advocacy Priorities;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities|Millennium Development Goals;CAP|MDG;Global Public Goods Priorities|Millennium Development Goals;GPG|MDG;Corporate Advocacy Priorities;CAP',
       'Corporate Advocacy Priorities|Millennium Development Goals;Corporate Advocacy Priorities|Millennium Development Goals;CAP|MDG;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities|Global Public Goods Priorities|Millennium Developmen

This column does not contain any relevance to out target variable and hence will be removed

## **impagency**

This column contains different agencies codes, and can be important in making decisions, thus imputing this column with 'mean'.

In [53]:
df_copy['impagency'].value_counts()

IIST                          4003
EET                            301
T                              290
EPATET                         188
SECETAIA                        98
                              ... 
AEEE                             1
ELECTICA                         1
L CU                             1
CS LA P                          1
RCPG                IPMENT       1
Name: impagency, Length: 3462, dtype: int64

In [62]:
# Imputing with mode, as IIST agency appeared 12 times more than any observation

df_copy['impagency'] = df_copy['impagency'].fillna(df_copy['impagency'].mode()[0])
print('impagency Missing values After Imputing:', df_copy['impagency'].isnull().sum())

impagency Missing values After Imputing: 0


## **borrower**
Similar to the 'impagency' column this column has the name of the borrowing agency

In [63]:
df_copy['borrower'].value_counts()

EET         3652
EPULIC      1640
IIST         749
I            274
T            258
            ... 
CAIC           1
LAL IT         1
IH             1
UEAU           1
ECCIECIA       1
Name: borrower, Length: 1639, dtype: int64